addons/binding/org.openhab.binding.smartmeter/src/main/java/org/openhab/binding/smartmeter/internal/MeterDevice.java

```diff
@@ -0,0 +1,254 @@
+import org.openhab.binding.smartmeter.connectors.IMeterReaderConnector;
+import org.openhab.binding.smartmeter.internal.helper.ProtocolMode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.reactivex.Flowable;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.schedulers.Schedulers;
+
+/**
+ *
+ * @author MatthiasS
+ *
+ * @param <T>
+ */
+@NonNullByDefault
+public abstract class MeterDevice<T> {
+
+    /**
+     * Controls wether the device info is logged to the OSGi console.
+     */
+    private boolean printMeterInfo;
+    /**
+     * Map of all values captured from the device during the read request.
+     */
+    private Map<String, MeterValue<?>> valueCache;
+    private byte @Nullable [] initMessage;
+    /**
+     * The id of the SML device from openHAB configuration.
+     */
+    private String deviceId;
+
+    /**
+     * Used to establish the device connection
+     */
+    IMeterReaderConnector<T> connector;
+    private List<MeterValueListener> valueChangeListeners;
+    private final static Logger logger = LoggerFactory.getLogger(MeterDevice.class);
+
+    public MeterDevice(Supplier<SerialPortManager> serialPortManagerSupplier, String deviceId, String serialPort,
+            byte @Nullable [] initMessage, int baudrate, int baudrateChangeDelay, ProtocolMode protocolMode) {
+        super();
+        this.deviceId = deviceId;
+        this.valueCache = new HashMap<String, MeterValue<?>>();
+        this.valueChangeListeners = new CopyOnWriteArrayList<>();
+        this.printMeterInfo = true;
+        this.connector = createConnector(serialPortManagerSupplier, serialPort, baudrate, baudrateChangeDelay,
+                protocolMode);
+    }
+
+    protected abstract IMeterReaderConnector<T> createConnector(Supplier<SerialPortManager> serialPortManagerSupplier,
+            String serialPort, int baudrate, int baudrateChangeDelay, ProtocolMode protocolMode);
+
+    /**
+     * Gets the configured deviceId.
+     *
+     * @return the id of the SmlDevice from openHAB configuration.
+     */
+    public String getDeviceId() {
+        return deviceId;
+    }
+
+    /**
+     * Returns the specified OBIS value if available.
+     *
+     * @param obis the OBIS code which value should be retrieved.
+     * @return the OBIS value as String if available - otherwise null.
+     */
+    @Nullable
+    public String getValue(String obisId) {
+        MeterValue<?> smlValue = getSmlValue(obisId);
+        if (smlValue != null) {
+            return smlValue.getValue();
+        }
+        return null;
+    }
+
+    /**
+     * Returns the specified OBIS value if available.
+     *
+     * @param obis the OBIS code which value should be retrieved.
+     * @return the OBIS value if available - otherwise null.
+     */
+    @SuppressWarnings("unchecked")
+    @Nullable
+    public <Q extends Quantity<Q>> MeterValue<Q> getSmlValue(String obisId) {
+
```

Comment:
Please remove such empty lines at the beginning of a method throughout your code!